markdown: Add Kroki integration to render diagrams from textual descriptions.

akshatdalton opened the pull request:
I've asked here: https://gitter.im/kroki-project/community# that if Kroki provides any library/API to get all the diagram types. Unfortunately, they don't have any such explicit API to provide this data. But they are in favour of adding this feature, so in future, we can use that instead of hardcoding these diagram types. (They have provided: https://kroki.io/health, which is not for this purpose, and I think we should not use this to get the diagram types.)

I've tried testing all the examples provided at: https://kroki.io/#try. I noticed that most of them are working, but a few of them are returning `Error 500: OK`. So I've asked the same here: https://gitter.im/kroki-project/community#.
Otherwise, this is working fine.

Previous discussions:
1. https://chat.zulip.org/#narrow/stream/49-developmenthelp/topic/pull.20docker.20images.20and.20run.20in.20zulip.20vagrant.20file
2. https://chat.zulip.org/#narrow/stream/137-feedback/topic/Mermaid.20Js.20renderer.20support

Fixes: #16825.

godlike786 committed:
tools: Add support to generate Kroki data.

In `kroki.json`, we have added `diagram_types` that
are currently available. See: https://kroki.io/#support.
Kroki doesn't provide any library or API to get all
these diagram types, but they may add this feature in
future then we no need to hard code these diagram types.

Co-authored-by: akshatdalton <akshat.dak@students.iiit.ac.in>

akshatdalton committed:
rendered_markdown: Add css for Kroki diagrams.

akshatdalton committed:
docs: Document Kroki data dependency.

akshatdalton commented:
https://github.com/zulip/zulip/blob/932c846454b5728658cb0e9ce05ad06947036849/static/js/composebox_typeahead.js#L614-L616

This will not show typeahead suggestions if 2 or more whitespace are added after the fence(` ``` `)

akshatdalton commented:
> I've tried testing all the examples provided at: https://kroki.io/#try. I noticed that most of them are working, but a few of them are returning `Error 500: OK`. So I've asked the same here: https://gitter.im/kroki-project/community#.
Otherwise, this is working fine.

This was a minor server problem which is now fixed, and everything is now working properly.

akshatdalton commented on zerver/lib/kroki.py:
```diff
@@ -0,0 +1,40 @@
+    )
+
+    try:
+        response = requests.post(url=settings.KROKI_URL, json=data, timeout=10)
```
`timeout` here is not decided.

timabbott commented on zproject/prod_settings_template.py:
```diff
@@ -692,3 +692,7 @@
+
+## We use this API provided by Kroki to render
+## diagrams from textual descriptions.
+KROKI_URL = "https://kroki.io/"
```
Let's leave adding this to `prod_settings_template.py` for the end; for this commit:
* You can set that URL in `dev_settings.py` and set `None` in `default_settings.py`, so that the feature is disabled except in the development environment, so that we can incrementally merge infrastructure commits without changing anything in production.
* We should call it `KROKI_SERVER_URL`, in analogy to `JITSI_SERVER_URL`.

timabbott commented on zerver/tests/test_markdown.py:
```diff
@@ -1156,6 +1158,58 @@ def test_fetch_tweet_data_settings_validation(self) -> None:
+        if not all(json_data.values()):
+            return MockPythonResponse("", 400)
+
+        return MockPythonResponse(args[1], 200)
```
We should use the `responses` module for mocking HTTP responses, rather than a custom system.

timabbott commented on zerver/tests/test_markdown.py:
````diff
@@ -1156,6 +1158,58 @@ def test_fetch_tweet_data_settings_validation(self) -> None:
+
+        # Basic test case.
+        msg = """\n```kroki graphviz\ndigraph G {\nHello->World\n}"""
+        html = """<?xml version="1.0" encoding="UTF-8" standalone="no"?>\n<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN"\n "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">\n<!-- Generated by graphviz version 2.44.0 ()\n -->\n<!-- Title: G Pages: 1 -->\n<svg width="74pt" height="116pt"\n viewBox="0.00 0.00 74.29 116.00" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlin ...
````
These are very long lines; could we read these from a fixture file in `zerver/fixtures/kroki` instead of having them inline here?

timabbott commented on zerver/tests/test_markdown.py:
```diff
@@ -1156,6 +1158,58 @@ def test_fetch_tweet_data_settings_validation(self) -> None:
+        expected = f"""<div class="kroki-block">{html}</div>"""
+        assert_kroki_diagrams(msg, expected, html)
+
+        # Kroki will not be able to generate html string unless diagram type is not provided.
```
I don't understand this comment.

akshatdalton replied:
Kroki cannot automatically guess the diagram type (when we don't provide it while making the POST request) -- hence it'll not generate an HTML string. (The expected output clarifies the same)

timabbott commented on zerver/lib/markdown/fenced_code.py:
```diff
@@ -364,6 +369,37 @@ def done(self) -> None:
+        if line.rstrip() == self.fence:
+            self.done()
+        else:
+            self.lines.append(line)
```
Does it makes sense to do a preparatory refactor PR to move functions like this into `BaseHandler` or perhaps a new `ZulipBaseHandler` class?  It feels like most of the code of these are duplicated.

timabbott commented on frontend_tests/node_tests/composebox_typeahead.js:
````diff
@@ -1302,6 +1305,8 @@ test("begins_typeahead", (override) => {
     assert_typeahead_equals("test\n~~~ p", lang_list);
     assert_typeahead_equals("test\n~~~  p", lang_list);
+    assert_typeahead_equals("```krok", lang_list);
+    assert_typeahead_equals("~~~kroki b", lang_list);
````
I'd like to see a test that more obviously shows the kroki full syntax being suggested.

timabbott commented on templates/zerver/help/code-blocks.md:
```diff
@@ -127,13 +127,14 @@ be happy to help you out.
-[Math blocks][math-block], [spoiler blocks][spoiler-block], and [quote
-blocks][quote-block] use similar fenced block syntax.
+[Math blocks][math-block], [Kroki diagram blocks][kroki-diagram-block],
+[spoiler blocks][spoiler-block], and [quote blocks][quote-block] use similar fenced block syntax.
```
Let's put the `Korki diagram blocks` option at the end of this list, since it's the most obscure.

timabbott commented:
Thanks for working on this @akshatdalton!  I merged the first commit 6143cb6e73a9092a1878a5bc9d07f862a16aab17, and posted a batch of comments that should let us starting merging a larger prefix of this PR.

godlike786 committed:
markdown: Add Kroki integration to render diagrams from textual descriptions.

We use API provided by Kroki to render diagrams from
textual descriptions. To render these diagrams, we need
to explicitly provide `diagram_source`, `diagram_type`
and `output_format` while making the POST request.
We are using `SVG` format type since all the diagram
sources are compatible with it.
See: https://kroki.io/#support.

Fixes: #16825.

Co-authored-by: akshatdalton <akshat.dak@students.iiit.ac.in>

godlike786 committed:
typeahead: Add typeahead support for Kroki diagram types.

This uses data generated in `kroki_data.json`.
Just like pygments typeahead, we will get typeahead
suggestions for Kroki diagram types when a fence in
opened.

Co-authored-by: akshatdalton <akshat.dak@students.iiit.ac.in>

akshatdalton committed:
help: Add documentation for Kroki integration.

akshatdalton commented:
Changes made in the current push:

- `KROKI_SERVER_URL` is set only in `dev_settings.py`. https://github.com/zulip/zulip/pull/18658#discussion_r644400378
- Used `responses` module instead of a custom system. https://github.com/zulip/zulip/pull/18658#discussion_r644400553
- Transferred the very long test cases to a new file, `zerver/fixtures/kroki.json`. https://github.com/zulip/zulip/pull/18658#discussion_r644400700
- Updated the comment in the codebase. https://github.com/zulip/zulip/pull/18658#discussion_r644400746
- Added `ZulipBaseHandler`, which de-duplicates most of the code. https://github.com/zulip/zulip/pull/18658#discussion_r644401135
- Added better test cases. https://github.com/zulip/zulip/pull/18658#discussion_r644401417
- Added the `Korki diagram blocks` option at the end of the list in the user-facing doc. https://github.com/zulip/zulip/pull/18658#discussion_r644401609

---

Many parts of the codebase use a custom system (like pointed in https://github.com/zulip/zulip/pull/18658#discussion_r644400553); I was trying to remove it and use the `responses` module -- but it looks that it is going to take much more time (+ in terms of understanding those codes). So I'll push the changes fixing them and also update the doc: https://zulip.readthedocs.io/en/latest/testing/testing-with-django.html#zulip-mocking-practices.

@timabbott, please review.

akshatdalton committed:
refactor: Add class `ZulipBaseHandler` to de-duplicate code.

akshatdalton commented:
(Just removed an unused function, added by me). Changes mentioned in the comment: https://github.com/zulip/zulip/pull/18658#issuecomment-855435021.

timabbott commented:
I merged the first commit as db1cf3b521785e52f02807190d2e33e7a5649758.  I think we can do a bit further cleanup by adding a `process_contents` flag so that the spoiler/quote ones don't need a copy of the `done()` function; want to try doing that?  Can just do that refactor in a new PR since it should now be independent of what we're doing here.

timabbott commented on zerver/tests/fixtures/kroki.json:
````diff
@@ -0,0 +1,12 @@
+    {
+        "name": "Kroki_basic_test_hello_world",
+        "input": "```kroki graphviz\ndigraph G {\nHello->World\n}",
+        "html_output": "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n<!-- Generated by graphviz version 2.44.0 ()\n -->\n<!-- Title: G Pages: 1 -->\n<svg width=\"74pt\" height=\"116pt\"\n viewBox=\"0.00 0.00 74.29 116.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"ht ...
````
This HTML looks like it's intended to be a complete webpage, not something embedded in another page.

Does `kroki` have an option to not do that?  If not, we should open an issue asking for them to do so.

akshatdalton replied:
I don't understand how this HTML is intended to be a complete webpage. We have the image in SVG format.

timabbott commented on zerver/lib/kroki.py:
```diff
@@ -0,0 +1,40 @@
+        diagram_source=text,
+        diagram_type=diagram_type,
+        output_format="SVG",
+    )
```
So I'm worried about the security implications of including arbitrary SVGs in a Zulip message.  I wonder if we want to make this integration instead upload the returned SVG as a Zulip file upload and then link to it like any other inline image, rather than putting it into our message content.

@andersk thoughts?  That process has somewhat involved infrastructure and is messy for message editing, so if there's a better way to restrict the XSS risk here, that'd be preferable.
